File: src/config/loader.ts
/**
 * Configuration loader - reads and validates config.json
 */

import fs from "fs";
import path from "path";
import { JsonConfig } from "./types.js";

/**
 * Load config.json from the current working directory
 * @returns JsonConfig if file exists, null if not found
 * @throws Error if file exists but is invalid
 */
export async function loadJsonConfig(): Promise<JsonConfig | null> {
  const configPath = path.resolve(process.cwd(), "config.json");

  // Check if file exists
  if (!fs.existsSync(configPath)) {
    return null;
  }

  try {
    const fileContent = fs.readFileSync(configPath, "utf-8");
    const config = JSON.parse(fileContent) as JsonConfig;

    // Validate required fields
    validateJsonConfig(config);

    return config;
  } catch (error) {
    if (error instanceof SyntaxError) {
      throw new Error(`Invalid JSON in config.json: ${error.message}`);
    }
    if (error instanceof Error) {
      throw error;
    }
    throw new Error("Failed to load config.json");
  }
}

/**
 * Validate JSON configuration structure
 * @throws Error if validation fails
 */
function validateJsonConfig(config: unknown): asserts config is JsonConfig {
  if (!config || typeof config !== "object") {
    throw new Error("Configuration must be a JSON object");
  }

  const cfg = config as Record<string, unknown>;

  // Validate dndBeyond section
  if (!cfg.dndBeyond || typeof cfg.dndBeyond !== "object") {
    throw new Error("Missing required section: dndBeyond");
  }
  const dnd = cfg.dndBeyond as Record<string, unknown>;
  if (!dnd.characterId || typeof dnd.characterId !== "string") {
    throw new Error("dndBeyond.characterId is required and must be a string");
  }
  if (!dnd.cobaltSession || typeof dnd.cobaltSession !== "string") {
    throw new Error("dndBeyond.cobaltSession is required and must be a string");
  }

  // Validate obs section
  if (!cfg.obs || typeof cfg.obs !== "object") {
    throw new Error("Missing required section: obs");
  }
  const obs = cfg.obs as Record<string, unknown>;
  if (!obs.websocketUrl || typeof obs.websocketUrl !== "string") {
    throw new Error("obs.websocketUrl is required and must be a string");
  }
  if (!obs.mode || !["image_swap", "visibility_toggle"].includes(obs.mode as string)) {
    throw new Error('obs.mode must be either "image_swap" or "visibility_toggle"');
  }

  // Validate mode-specific requirements
  if (obs.mode === "visibility_toggle") {
    if (!obs.sceneName || typeof obs.sceneName !== "string") {
      throw new Error("obs.sceneName is required for visibility_toggle mode");
    }
  } else if (obs.mode === "image_swap") {
    if (!obs.sourceName || typeof obs.sourceName !== "string") {
      throw new Error("obs.sourceName is required for image_swap mode");
    }
    if (!obs.images || typeof obs.images !== "object") {
      throw new Error("obs.images is required for image_swap mode");
    }
    const images = obs.images as Record<string, unknown>;
    const requiredImages = ["healthy", "scratched", "bloodied", "dying", "unconscious"];
    for (const img of requiredImages) {
      if (!images[img] || typeof images[img] !== "string") {
        throw new Error(`obs.images.${img} is required for image_swap mode`);
      }
    }
  }

  // Validate polling section
  if (!cfg.polling || typeof cfg.polling !== "object") {
    throw new Error("Missing required section: polling");
  }
  const polling = cfg.polling as Record<string, unknown>;
  if (typeof polling.intervalMs !== "number" || polling.intervalMs < 1000) {
    throw new Error("polling.intervalMs must be a number >= 1000");
  }

  // Validate gameLog section if present
  if (cfg.gameLog && typeof cfg.gameLog === "object") {
    const gameLog = cfg.gameLog as Record<string, unknown>;
    if (gameLog.enabled === true) {
      if (!gameLog.gameId || typeof gameLog.gameId !== "string") {
        throw new Error("gameLog.gameId is required when gameLog.enabled is true");
      }
      if (!gameLog.userId || typeof gameLog.userId !== "string") {
        throw new Error("gameLog.userId is required when gameLog.enabled is true");
      }
    }
  }
}

/**
 * Save configuration to config.json
 */
export async function saveJsonConfig(config: JsonConfig): Promise<void> {
  const configPath = path.resolve(process.cwd(), "config.json");

  try {
    const jsonString = JSON.stringify(config, null, 2);
    fs.writeFileSync(configPath, jsonString, "utf-8");
  } catch (error) {
    throw new Error(
      `Failed to save config.json: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}

/**
 * Check if config.json exists in current working directory
 */
export function configExists(): boolean {
  const configPath = path.resolve(process.cwd(), "config.json");
  return fs.existsSync(configPath);
}

File: rust-src/config/loader.rs
use anyhow::{bail, Context, Result};
use std::path::Path;

use super::types::JsonConfig;

/// Load config.json from the given path
pub async fn load_json_config(config_path: &Path) -> Result<Option<JsonConfig>> {
    if !config_path.exists() {
        return Ok(None);
    }

    let content =
        tokio::fs::read_to_string(config_path)
            .await
            .context("Failed to read config.json")?;

    let config: JsonConfig =
        serde_json::from_str(&content).context("Invalid JSON in config.json")?;

    validate_json_config(&config)?;
    Ok(Some(config))
}

/// Validate JSON configuration structure
fn validate_json_config(config: &JsonConfig) -> Result<()> {
    // Validate dndBeyond section
    if config.dnd_beyond.character_id.is_empty() {
        bail!("dndBeyond.characterId is required and must be a string");
    }
    if config.dnd_beyond.cobalt_session.is_empty() {
        bail!("dndBeyond.cobaltSession is required and must be a string");
    }

    // Validate obs section
    if config.obs.websocket_url.is_empty() {
        bail!("obs.websocketUrl is required and must be a string");
    }
    if config.obs.mode != "image_swap" && config.obs.mode != "visibility_toggle" {
        bail!("obs.mode must be either \"image_swap\" or \"visibility_toggle\"");
    }

    // Validate mode-specific requirements
    if config.obs.mode == "visibility_toggle" {
        if config.obs.scene_name.as_ref().map_or(true, |s| s.is_empty()) {
            bail!("obs.sceneName is required for visibility_toggle mode");
        }
    } else if config.obs.mode == "image_swap" {
        if config.obs.source_name.as_ref().map_or(true, |s| s.is_empty()) {
            bail!("obs.sourceName is required for image_swap mode");
        }
        if config.obs.images.is_none() {
            bail!("obs.images is required for image_swap mode");
        }
    }

    // Validate polling section
    if config.polling.interval_ms < 1000 {
        bail!("polling.intervalMs must be >= 1000");
    }

    // Validate gameLog section if present and enabled
    if let Some(ref game_log) = config.game_log {
        if game_log.enabled {
            if game_log.game_id.as_ref().map_or(true, |s| s.is_empty()) {
                bail!("gameLog.gameId is required when gameLog.enabled is true");
            }
            if game_log.user_id.as_ref().map_or(true, |s| s.is_empty()) {
                bail!("gameLog.userId is required when gameLog.enabled is true");
            }
        }
    }

    Ok(())
}

/// Save configuration to config.json
pub async fn save_json_config(config: &JsonConfig, config_path: &Path) -> Result<()> {
    let json_string =
        serde_json::to_string_pretty(config).context("Failed to serialize config")?;
    tokio::fs::write(config_path, json_string)
        .await
        .context("Failed to write config.json")?;
    Ok(())
}

/// Check if config.json exists
pub fn config_exists(config_path: &Path) -> bool {
    config_path.exists()
}
